crit/decoder.c

```diff
@@ -0,0 +1,125 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdint.h>
+#include <fcntl.h>
+
+#ifdef CONFIG_HAS_FLOG
+typedef struct {
+	unsigned int	magic;
+	unsigned int	size;
+	unsigned int	nargs;
+	unsigned int	mask;
+	long		fmt;
+	long		args[0];
+} flog_msg_t;
+
+#define MAGIC 0xABCDABCD
+
+#define BUF_SIZE (1<<20)
+static char _mbuf[BUF_SIZE];
+static char *mbuf = _mbuf;
+static char *fbuf;
+static uint64_t fsize;
+static uint64_t mbuf_size = sizeof(_mbuf);
+
+#define LOG_BUF_LEN		(8*1024)
+static char buffer[LOG_BUF_LEN];
+static char buf_off = 0;
+
+int decode_all(int fdin, int fdout)
+{
+	flog_msg_t *m = (void *)mbuf;
+	void *values[34], *b;
+	size_t i, ret;
+	char *fmt;
+	int size, n;
+
+	while (1) {
+		ret = read(fdin, mbuf, sizeof(m));
+
+		if (ret == 0)
+			break;
+		if (ret < 0) {
+			fprintf(stderr, "Unable to read a message: %m");
+			return -1;
+		}
+		if (m->magic != MAGIC) {
+			lseek(fdin, -sizeof(ret), SEEK_CUR);
+			read(fdin, b, 1);
```

Comment:
You need to check return codes of ALL functions.

Reply:
I will make this change and return the appropriate error messages for the same.